### PR TITLE
[fx] Add common node in model linearize

### PR DESCRIPTION
**Pull request by Cypher30**
# What's New

We were inspired by the common node idea in [merak](https://arxiv.org/abs/2206.04959), and modify the naive process of defining a common node in [merak's source code](https://github.com/HPDL-Group/Merak/blob/91864243b21ccfef1d58746623e3a6cebdc630ae/Merak/autoshard/convert.py#L134). Now we are able to address the attention mask problem in hugginface BERT model, separating each attention block and MLP in the encoder.

Moreover, I also fixed the problem that our solver might repeatedly count the memory for combination like "BN-RELU", as the current linearize strategy will generate more isolated in-place RELU node in the linearized node list.

### CODE REVIEW

File: colossalai/fx/passes/algorithms/ckpt_solver_rotor.py [hunk 1]
```diff
@@ -7,6 +7,7 @@
 from .utils import *
 from colossalai.fx.profiler import profile_function, profile_module
 from colossalai.fx.passes.meta_info_prop import MetaInfoProp
+import pdb
```
Comment (super-dainiu) [inline colossalai/fx/passes/algorithms/ckpt_solver_rotor.py:10]: remove this

File: colossalai/fx/passes/algorithms/linearize.py [hunk 2]
```diff
@@ -2,11 +2,12 @@
 from torch.fx import GraphModule, Node
 
 
-def linearize(gm: GraphModule) -> List[List[Node]]:
+def linearize(gm: GraphModule, cnode: List[str] = None) -> List[List[Node]]:
```
Comment (super-dainiu) [inline colossalai/fx/passes/algorithms/linearize.py:5]: where do you get the cnode?

File: colossalai/fx/passes/algorithms/ckpt_solver_rotor.py (hunk 1 above)
Comment (Cypher30) [reply colossalai/fx/passes/algorithms/ckpt_solver_rotor.py:10]: okay, forget lol

File: colossalai/fx/passes/algorithms/linearize.py (hunk 2 above)
Comment (Cypher30) [reply colossalai/fx/passes/algorithms/linearize.py:5]: yes, it should be added to solver

Comment (super-dainiu) [reply colossalai/fx/passes/algorithms/linearize.py:5]: I mean how do you get the list of cnode?

Comment (Cypher30) [reply colossalai/fx/passes/algorithms/linearize.py:5]: The first list of common nodes should be input by users, and the other common nodes will be propagated in the following codes.

**Review by super-dainiu — APPROVED** [review]

**Pull request merged**